File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/KitchenRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00004\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0002\b\u0002\n\u0002\u0010!\n\u0000\u0018\u00002\u00020\u0001B\u0017\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\u0002\u0010\u0006J\u001c\u0010\u0007\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\n0\t0\b2\u0006\u0010\u000b\u001a\u00020\fH\u0016J\u0014\u0010\r\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\n0\t0\bH\u0016J\u001a\u0010\u000e\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\n0\u000f0\t0\bH\u0016R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0010"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/KitchenRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/KitchenRepository;", "kitchenDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/KitchenDao;", "ioDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/local/dao/KitchenDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getKitchenId", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/utils/Resource;", "Lcom/bits/bee/bpmc/domain/model/Kitchen;", "id", "", "getLastKitchen", "readKitchen", "", "app_debug"})
public final class KitchenRepositoryImpl implements com.bits.bee.bpmc.domain.repository.KitchenRepository {
    private final com.bits.bee.bpmc.data.data_source.local.dao.KitchenDao kitchenDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher ioDispatcher = null;
    
    @javax.inject.Inject()
    public KitchenRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.KitchenDao kitchenDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher ioDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.Kitchen>> getKitchenId(int id) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.Kitchen>> getLastKitchen() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Kitchen>>> readKitchen() {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/KitchenRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.KitchenDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.KitchenResponse
import com.bits.bee.bpmc.domain.mapper.KitchenDataMapper
import com.bits.bee.bpmc.domain.model.Kitchen
import com.bits.bee.bpmc.domain.repository.KitchenRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import javax.inject.Inject

class KitchenRepositoryImpl @Inject constructor(
    private val kitchenDao: KitchenDao,
    private val apiUtils: ApiUtils,
    private val ioDispatcher: CoroutineDispatcher
): KitchenRepository {

    override fun getLatestKitchen(): Flow<Resource<List<Kitchen>>> {
        return object : NetworkDatabaseBoundResource<List<Kitchen>, KitchenResponse>(){
            override suspend fun loadFormDB(): List<Kitchen> {
                return kitchenDao.read().map { KitchenDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<Kitchen>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<KitchenResponse>> {
                return apiUtils.getKitchenApiService().getKitchen()
            }

            override suspend fun saveCallResult(data: KitchenResponse) {
                kitchenDao.insertBulk(data.kitchenModels.map { KitchenDataMapper.fromNetworkToDb(it) })
            }
        }.getAsFlow()
    }

    override fun getKitchenId(id: Int): Flow<Kitchen> {
        return flow {
            val data = kitchenDao.getByKitchenId(id)
            emit(KitchenDataMapper.fromDbToDomain(data))
        }.flowOn(ioDispatcher)
    }

    override fun getLastKitchen(): Flow<Kitchen> {
        return flow {
            val data = kitchenDao.getLastKitchen()
            emit(KitchenDataMapper.fromDbToDomain(data))
        }.flowOn(ioDispatcher)
    }

    override fun readKitchen(): Flow<MutableList<Kitchen>> {
        return flow {
            val data = kitchenDao.read().map { KitchenDataMapper.fromDbToDomain(it) }
            emit(data.toMutableList())
        }.flowOn(ioDispatcher)
    }
}